### PR TITLE
Update hardware sim and toast workflows.

### PR DESCRIPTION
**Pull request by tskisner**
* Use new xi/eta/gamma support in toast when simulating detector layout.

* Fix projected LAT rotation on the sky, and also the tube positions.

* Add LAT co-rotation options when making focalplane plots.

* Support focalplane plots in both Xi / Eta and X / Y coordinates.

* Cache detector beams in SimSSO and SimSource, in case we have the same beam for many detectors.

* Small tweaks to get_wafer_offsets arguments

* Add a new toast workflow that just loads data and reduces it.

### CODE REVIEW

**Comment by tskisner** [comment]
This PR against the toast3 branch adds fixes for the detector focalplane layout needed to produce consistent LAT tube geometry and rotation.  One change is that simulating the hardware model is now done with a `sim_nominal()` function, rather than the poorly named `get_example()`.  The returned model has no detectors (only the telescope / tube / wafer / band information).  Then one of several functions can be called to simulate detector locations.  The default one uses the toast geometry layout helper functions, but there is a stub for a "better" one using optics data.

One consequence of this is that to do the initial "full" instrument detector sim and generate a TOML / JSON file, toast is required.  After that file exists, it can be modified, trimmed, plotted, etc without the toast sub-module.  So I would advocate that we automate the sotodlib releases to append one of these instrument files to the release page (or some other place).

I am running a SAT test now.

File: workflows/toast_so_sim.py [hunk 1]
```diff
@@ -221,44 +219,77 @@ def job_create(config, jobargs, telescope, schedule, comm):
...
+    ops.weights_azel.hwp_angle = ops.sim_ground.hwp_angle
+
+    ops.weights_radec.detector_pointing = ops.det_pointing_radec
+    ops.weights_radec.hwp_angle = ops.sim_ground.hwp_angle
+
+    # Select Pixelization and weights for solve and final binning
+
+    if ops.pixels_wcs_azel.enabled:
+        args.pixels_solve = ops.pixels_wcs_azel
+        args.weights_solve = ops.weights_azel
```
Comment (keskitalo) [inline workflows/toast_so_sim.py:274]: I'm confused. Where is `weights_solve` added to the `args` namespace?

File: tests/test_hardware.py [hunk 2]
```diff
@@ -7,19 +7,24 @@
...
+from sotodlib.vis_hardware import plot_detectors
 
-from sotodlib.sim_hardware import get_example
+from ._helpers import create_outdir, mpi_multi
 
-from sotodlib.sim_hardware import (sim_wafer_detectors,
-                                   sim_telescope_detectors)
 
-from sotodlib.vis_hardware import plot_detectors
+toast_available = None
```
Comment (keskitalo) [inline tests/test_hardware.py:19]: I think we have discussed this import guard before. Since you set `toast_available` to `None` and immediately check its value, the test is redundant and can be omitted. Furthermore, because of Python scoping rules, you don't even need to declare `toast_available` before the `try-except` clause.

File: docs/toast.rst [hunk 3]
```diff
@@ -104,10 +104,14 @@ When running workflows in batch mode, we typically use a slightly different
 structure from a notebook.  There are helper functions in toast to parse
 operator options from the commandline and / or a config file, and to automatically
 create named instances of those operators for use in workflow.  There is
-a large-scale workflow in ``sotodlib/workflows/toast_so_sim.py``.  This
+a large-scale workflow in ``sotodlib/workflows/toast_so_map.py``.  This
```
Comment (keskitalo) [inline docs/toast.rst:107]: I think you mixed `toast_so_map.py` and `toast_so_sim.py` on line 111.

File: sotodlib/toast/ops/sim_sso.py [hunk 4]
```diff
@@ -131,6 +121,11 @@ def _check_detector_pointing(self, proposal):
 
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
+        # Store of per-detector beam properties.  Eventually we could modify the
```
Comment (keskitalo) [inline sotodlib/toast/ops/sim_sso.py:124]: I really appreciate these helpful comments.

File: sotodlib/toast/ops/sim_sso.py [hunk 5]
```diff
@@ -278,62 +292,96 @@ def _get_sso_position(self, data, sso, times, observer):
...
             x = (az - sso_az.to_value(u.rad)) * np.cos(el)
             y = el - sso_el.to_value(u.rad)
-            r = np.sqrt(x ** 2 + y ** 2)
+            r = np.sqrt(x**2 + y**2)
+
             good = r < radius
             sig = beam(x[good], y[good], grid=False)
             signal[good] += sig
 
+            # if idet == 0:
```
Comment (keskitalo) [inline sotodlib/toast/ops/sim_sso.py:363]: Did you intend to drop these debug statements?

File: sotodlib/toast/ops/sim_sso.py [hunk 6]
```diff
@@ -363,3 +411,137 @@ def _provides(self):
 
     def _accelerators(self):
         return list()
+
```
Comment (keskitalo) [inline sotodlib/toast/ops/sim_sso.py:414]: And these lines?

**Review by keskitalo — APPROVED** [review]
Thanks a ton on your thorough work on the focalplane simulation and the co-rotator!

I only have minor comments. Please see if they merit any action.

**Review by smsimon — APPROVED** [review]
There's been a desire from the instrument side to skip a detector number on the pin/slot (i.e. killed pixels) so that if pin/slot locations change (e.g. in Advanced SO) or we kill pixels that are dead in the model, the detector numbers don't change. Now might be a good time to implement that in sim_focalplane.py. I _think_ you just have to change doff to px in line 354 (or something similar) though we should check if it has any other implications.

File: sotodlib/toast/ops/corotator.py [hunk 7]
```diff
@@ -86,35 +98,46 @@ def _exec(self, data, detectors=None, **kwargs):
             if obs.telescope.name != "LAT":
                 continue
 
+            el_rad = obs.shared[self.elevation]
+
+            if self.corotator_angle not in obs.shared:
```
[the remark this answers is not in the document]
Comment (tskisner) [reply sotodlib/toast/ops/corotator.py:103]: The `ensure()` method is for detector data (https://github.com/hpc4cmb/toast/blob/ee47b337eccf57f5911893e1b1b7021219ff5fe0/src/toast/observation_data.py#L608) where it can re-use the existing detector data buffer by re-assigning it to different detector sets.

File: workflows/toast_so_sim.py (hunk 1 above)
Comment (tskisner) [reply workflows/toast_so_sim.py:274]: This is done in the new `select_pointing()` [function here](https://github.com/simonsobs/sotodlib/blob/d052fb851f1ccd565be3c1ae1e93deb4b0d8fed4/workflows/toast_so_sim.py#L272-L281)  Many of these workflow helper functions (which deal with runtime options configuring multiple operators) can move to within the toast submodule.  However, I wanted to do that in a future, smaller PR.

File: tests/test_hardware.py (hunk 2 above)
Comment (tskisner) [reply tests/test_hardware.py]: Good point, I removed that.  Also, in the `toast3_data` PR, I moved this check into the `_helpers.py` file for use in multiple tests.

File: docs/toast.rst (hunk 3 above)
Comment (tskisner) [reply docs/toast.rst]: Yes, clarified and fixed.

File: sotodlib/toast/ops/sim_sso.py (hunk 5 above)
Comment (tskisner) [reply sotodlib/toast/ops/sim_sso.py]: I removed the commented out debugging statements from the operator itself, but left the commented out plotting function at the bottom of the source.  We could delete that later or move to some other source if it is useful.

File: workflows/toast_so_sim.py (hunk 1 above)
Comment (keskitalo) [reply workflows/toast_so_sim.py:274]: I see. Sort of like a `workflow_tools` package? I'm very much in favor of that!

I was thrown off because symbols in the `args` namespace are usually declared via `argparse` but now we are expanding it to include symbols generated as the workflow progresses. That is fine with me but it is a slippery slope into using it as a vehicle for all information flow in the workflow. Perhaps we want to call it something else than `args` if we expand its meaning?

Comment (tskisner) [reply workflows/toast_so_sim.py:274]: Ok, I see your point on abusing the `args` namespace.  I propose storing these kinds of workflow variables in the `job` namespace that we create when loading configs / options and instantiating the operators and templates.  For the high-level set of functions in the toast submodule, these should be helper functions that work on configuring many operators together.  The old "`pipeline_tools`" code was basically trying to externally configure individual operators.  Now that operators have all their configuration built-in, the kinds of things we would put in `sotodlib.toast.workflow` would be things like:

- Return a list of all nominal simulation operators and their starting defaults
- Return a list of all nominal reduction operators and their starting defaults
- Select the pointing / weights being used depending on which operators are enabled at runtime
- Function to "save" or "load" data using whichever operator combination has been enabled at runtime

The `simulate_data()` and `reduce_data()` functions in the workflows are kind of specific to the workflow, but I guess we could have a `simulate_nominal_data()` or similar in the workflow submodule.  There is a balance between reducing boilerplate / duplication in the script and hiding what the script is doing.  If a workflow has a main() that just calls 3 functions, it is really hard to follow what it is actually doing.  On the other hand, perhaps "following what is going on" is better left to interactive notebooks, and these workflows should just be about running things while reducing code duplication across scripts.

I'll fix the use of the args namespace here, and we can discuss further the workflow submodule on the next PR.

File: sotodlib/toast/ops/corotator.py [hunk 8]
```diff
@@ -86,35 +98,46 @@ def _exec(self, data, detectors=None, **kwargs):
...
+                    self.corotator_angle,
+                    shape=(obs.n_local_samples,),
+                    dtype=np.float64,
+                )
+
             if self.corotate_lat:
-                obs["corotator_angle"] = -obs["scan_el"]
+                # We are corotating the receiver.  We set this to a fixed angle
+                # based on the nominal elevation for the scan.
+                corot = obs["scan_el"].to_value(u.radian) - LAT_COROTATOR_OFFSET.to_value(u.radian)
```
[the remark this answers is not in the document]
Comment (tskisner) [reply sotodlib/toast/ops/corotator.py]: This is now changed to use the mean boresight elevation if the target / intended "scan_el" is not defined.

File: workflows/toast_so_sim.py (hunk 1 above)
Comment (tskisner) [reply workflows/toast_so_sim.py]: This is now fixed

**Comment by tskisner** [comment]
The changes in this PR are now rebased against the toast3 branch and have changes to make the code compatible with the latest toast-3 pre-release.  The `toast3_bcp` branch of pwg-scripts is now also compatible with this code, and the drone sim, Jupiter SAT, and Jupiter LAT examples are working:
drone_10_08
out_f150_i1_Jupiter
out_f030_w42_Jupiter

I will merge this PR after the upstream toast-3.0.0a5 wheels are available.

**Comment by tskisner** [comment]
Upstream wheels of toast that support this are now on PyPI.  Merging.
